### ISSUE TITLE
Do you really mean it to have a viral license?

### ISSUE THREAD

**Issue by KOLANICH** [dslackw/colored#7]

**Comment by dslackw** [comment]
Hi @KOLANICH, 
I'm not very familiar with the licences. What is the difference?

**Comment by KOLANICH** [comment]
A viral license requires everything using anything under that license have the same license. GPL is a viral one. In your case it means that every program or library directly importing and using your library is required to be under GPL. By induction it can be proven that any program having your library in its dependency graph is required to be under GPL.

It's bad because if it's under GPL it is not allowed to be under public-domain-like licenses like CC0 or Unlicense, permissive ones like Apache, BSD or MIT or any other license other than GPLv3.

Of course I can recreate the lib myself, but since I don't really need it, I prefered to ask first if this effect was intended.

If you want to relicense it you must get the consent from all the contributors, othervise it won't take any effect and you may be sues by that contributors or even Free Software Foundation.

**Comment by dopsi** [comment]
@dslackw If you want to keep a GNU license, there is the
[LGPL](https://www.gnu.org/licenses/why-not-lgpl.en.html) (see also [this
guide](https://www.gnu.org/licenses/license-recommendations.html#libraries)
explains the FSF's recommendations for licensing), which only
takes away the requirement of providing the source code of any **distributed** project using
your library (this requirement does not mean the source code has to be
published on the internet, but it could also be physically provided on
any other data transfer medium).

A good comparison of licenses is also available
[here](https://choosealicense.com/licenses/). However, as stated by
@KOLANICH, you can only relicense **your** code (or code you have
written approval to do so).

The question is whether or not you mind having your library used in proprietary
software, which @KOLANICH would like to do, but cannot as long as it is
licensed under GPL.

**Comment by KOLANICH** [comment]
Not quite. LGPL is still bad because to reuse the code you will still have to take parts of the original source code, at least names. For Rust, C, C++ and any other language having any preprocessor, it's a big issue because preprocessor directives are expanded, which causes embedding the library sources into project sources which causes requirement to have that project under the same license. So LGPL is an extremily bad choice for any lib using such languages, unless you are a copyright troll. It is suitable if there is an extemption or if the headers are in public domain but the implementation (which is not included int oprogram, just linked against) is lgpl. For other languages it's a lesser issue, but still an issue, since the dev must remember not to use the names larger than 2 lines. And the worst thing is that what is  "2 lines" is not defined. Are they automatically word-wrapped lines and at what screen resolution and in what editor and at what font and in what viewport area, or lines in meaning "pieces of text ended with \n", and what if someone then modified the lib (he is allowed to do so) and removed these \ns to fit the macros into a single line, or maybe it is "just 80 characters"? It's pretty ambiguos.

**Comment by dopsi** [comment]
@KOLANICH LGPLv3 section 4 make it very clear what the requirements are for
distributing binary copies of the program using an LGPL library:

* make it clear that the library is used and that the library is covered
under LGPL
* distribute a copy of the LGPL along the project
* if you display a copyright notice during the execution, display the
LGPL along with it
* if you link the library statically, provide the minimal source code
and object code to be able to link a different version (modified or
not) of the library with it
* if you dynamically link your library with the binary code, make sure
the program will be able to run with a modified library
* in both cases, you have to provide the instructions how to make it
work with the modified library

What makes you say there are any restrictions regarding the preprocessor
code for compiled languagues ?

What do you mean by this mumbo-jumbo:
 
> since the dev must remember not to use
> the names larger than 2 lines. And the worst thing is that what is "2 lines" is
> not defined. Are they automatically word-wrapped lines and at what screen
> resolution and in what editor and at what font and in what viewport area, or
> lines in meaning "pieces of text ended with \n", and what if someone then
> modified the lib (he is allowed to do so) and removed these \ns to fit the
> macros into a single line, or maybe it is "just 80 characters"? It's pretty
> ambiguos.

The LGPL states that you may incorporate any material from the headers without restrictions (inline functions and templates up to 10 lines) if you give proper notice and distribute a copy of the LGPL along with it. (In code, a line is defined by all the characters between `\n` or `\r` (or a combination of both)).

**Comment by KOLANICH** [comment]
>any material from the headers without restrictions (inline functions and templates up to 10 lines)

1 it's not "without restrictions"
2 thank you for the corrections, I have fixed the previous message
3 it's hard for developers not to violate this rule. A copyright troll can intentionally create a library, publish it under LGPL, but create a macrodef or inline function larger than 10 lines without which it is impossible to use the lib. Then he waits for the devs ignorant about this trap in the LGPL and then sues them.

> In code, a line is defined by all the characters between \n or \r (or a combination of both)).

It's not defined in the license. So it's defined by lawers. And in your definition it's unenforceable: everyone can fork a library under LGPL and minify the source, putting thousands of "your" (in your definition) lines into a single "your" line. Taking it into account lawers can decide (I'm not a lawer, this is only a hypothesis) for example that a line is just `n` characters because the style guide requires to wrap the lines by `n` characters, or that the lines are the lines in the original code written by its original author.

**Comment by tleonhardt** [comment]
I love your module, but I hate your licensing of it.  Making it GPL will severely restrict its use by major corporations who will be scared away by the viral licensing.

If you don't have any strong preferences in regards to licensing, pretty much **anything** would be better than GPL.  MIT, BSD, and Apache are all great.  LGPL is tolerable, though flawed for various reasons as you have mentioned.

I have some ideas for improvements to contribute to this module that I would like to experiment with, but I would prefer to not spend my time contributing to GPL code that I would never be able to use in my professional career.

Its your module, but I thought I would throw my 2 cents in since this Issue is currently open.

**Comment by tleonhardt** [comment]
@jwarner112 @brrzap @aluttik
Would any of the contributors object changing to an MIT license?

The project has moved to GitLab and the associated issue there is:
https://gitlab.com/dslackw/colored/issues/7

**Comment by aluttik** [comment]
@tleonhardt no objection here
